### PR TITLE
a11y traversal: sort locally; use new sorting algorithm

### LINKED ISSUE #12187: a11y traversal order top left to bottom right not always giving great results
Traversal order in this screen shot is currently:

1. Back button
2. Shopping cart button
3. Three dots button
4. "SHRINE" title

I would expect SHRINE to come after the back button. However, the shipping cart button's bounding box has its top higher than the SHRINE title and is therefore traversed first in the current top left to bottom right approach.

img_0011

### LINKED ISSUE #14375: BottomNavigationBar is not accessible

### LINKED ISSUE #14570: Semantics traversal order needs to be implemented on iOS engine.
The Semantics traversal order API has been implemented in the framework and on Android, but the iOS engine work still needs to be completed.

The engine now receives, for each semantics node, a "`nextNodeId`" which is the next node in the traversal order.  On Android this calls [`setTraversalBefore`](https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo.html#setTraversalBefore(android.view.View)) to indicate the order, but on iOS there isn't a similar facility.

The place in the code that orders accessibility nodes is [here](https://github.com/flutter/engine/blob/c210d62a6c38df62c2f3b01053b50603f4fd74dd/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm#L431).  It currently orders them geometrically (upper left to lower right), but we don't want that (it doesn't respect directionality, or the custom ordering API in the framework).

In addition, it orders them only on the basis of ordering the children of each node, and not globally: the custom ordering is completely orthogonal to the widget hierarchy, and could order any semantics node after any other.

I think that an approach similar to the one given in [stackoverflow here](https://stackoverflow.com/questions/13279498/change-order-of-read-items-with-voiceover) is a possible solution, but I don't know enough iOS development to know for sure.

This is part of implementing #12187

### PR DESCRIPTION
**Pull request by yjbanov**
- Remove global sorting that's not supported by iOS. Instead it sorts locally using a new algorithm [illustrated here](https://docs.google.com/presentation/d/1z-sLVogRPA8R0PHf1R-aMylS1UVQP3KQi9vO6apV4g0).
- RTL
- Remove `nextNodeId`/`previousNodeId`

Fixes https://github.com/flutter/flutter/issues/12187
Fixes https://github.com/flutter/flutter/issues/14375
Fixes https://github.com/flutter/flutter/issues/14570 (together with https://github.com/flutter/engine/pull/4937)

TODO:

- [x] remove traces of geometric sort order (e.g. the enum option)
- [ ] pass hit test order so it can be applied on Android (iOS doesn't support it; this requires an engine-side change)

/cc @goderbauer @gspencergoog

### CODE REVIEW

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 1]
```diff
@@ -1372,6 +1370,55 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
...
+    }
+
+    // Do not sort groups with null sort keys. List.sort does not guarantee
+    // a stable sort order.
+    if (lastSortKey != null) {
+      sortNodes.sort();
+    }
+    everythingSorted.addAll(sortNodes);
+
+    return everythingSorted
```
Comment (goderbauer) [inline packages/flutter/lib/src/semantics/semantics.dart]: Looks like everythingSorted could just be a List<SemanticsNode> from the beginning to avoid this conversion?

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 2]
```diff
@@ -1496,92 +1543,271 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
   }
 }
 
+/// An edge of a box, such as top, bottom, left or right, used to compute
+/// [SemanticsNode]s that overlap vertically or horizontally.
```
Comment (goderbauer) [inline packages/flutter/lib/src/semantics/semantics.dart]: Maybe document: For computing horizontal overlap in an LTR setting we create two _BoxEdge objects for each Semantics Node: one representing the left edge (marked with isStart: true) and one for the right edge (with isStart: false). Similarly, for vertical overlap you also create two objects for each SemanticsNode, one for the top and one for the bottom edge.

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 3]
```diff
@@ -1496,92 +1543,271 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
...
+  /// True if the edge comes before the seconds edge along the traversal
+  /// direction, and false otherwise.
+  ///
+  /// This field is never null.
+  ///
+  /// For example, in LTR traversal the left edge's [isStart] is set to true,
+  /// the right edge's [isStart] is set to false. When considering vertical
+  /// ordering of boxes, the top edge is the start edge, and the bottom edge is
+  /// the end edge.
+  final bool isStart;
```
Comment (goderbauer) [inline packages/flutter/lib/src/semantics/semantics.dart]: I believe in other parts of the flutter framework we call this the leading edge. Maybe rename for consistency?

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 4]
```diff
@@ -1496,92 +1543,271 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
...
+
+    final List<SemanticsNode> result = <SemanticsNode>[];
+    for (_SemanticsSortGroup group in horizontalGroups) {
+      final List<SemanticsNode> sortedKnotNodes = group.sortedWithinKnot();
+      result.addAll(sortedKnotNodes);
+    }
+    return result;
+  }
+
+  /// Sorts [nodes] where nodes intersect both vertically and horizontally.
```
Comment (goderbauer) [inline packages/flutter/lib/src/semantics/semantics.dart]: Slight confusion: Doesn't this method also get called for lists of length one that don't have overlap per definition?

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 5]
```diff
@@ -1496,92 +1543,271 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
...
+  /// This method constructs a graph, where vertices are [SemanticsNode]s and
+  /// edges are "traversed before" relation between pairs of nodes. The sort
+  /// order is the topological sorting of the graph, with the original order of
+  /// [nodes] used as the tie breaker.
+  ///
+  /// Whether a node is traversed before another node is determined by the
+  /// vector that connects the two nodes' centers. If the vector "points to the
+  /// right or down", defined as the [Offset.direction] being between `-pi/4`
+  /// and `3*pi/4`), then the semantics node whose center is at the end of the
+  /// vector is said to be traversed after.
```
Comment (goderbauer) [inline packages/flutter/lib/src/semantics/semantics.dart]: bonus points: would be great to link to the graphic from you slides here.

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 6]
```diff
@@ -1496,92 +1543,271 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
...
+        return;
+      }
+      visitedIds.add(id);
+      if (edges.containsKey(id)) {
+        search(edges[id]);
+      }
+      sortedIds.add(id);
+    }
+
+    startNodes.map((SemanticsNode node) => node.id).forEach(search);
```
Comment (goderbauer) [inline packages/flutter/lib/src/semantics/semantics.dart]: to avoid all this mapping, can search not operate on semanticsnodes and sortedIds be a List<SemanticNode>?

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 7]
```diff
@@ -1372,6 +1370,55 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
     _dirty = false;
   }
 
+  /// Builds a new list made of [_children] sorted in semantic traversal order.
+  List<SemanticsNode> _childrenInTraversalOrder() {
+    final List<SemanticsNode> childrenInDefaultOrder = _childrenInDefaultOrder(_children);
```
Comment (goderbauer) [inline packages/flutter/lib/src/semantics/semantics.dart]: do we have to do all this work to bring nodes, that specify a sort key, into default sort order only to re-sort them then according to their sort key?

File: packages/flutter/test/material/bottom_navigation_bar_test.dart [hunk 8]
```diff
@@ -512,32 +512,36 @@ void main() {
       ),
     );
 
-    // TODO(goderbauer): traversal order is incorrect, https://github.com/flutter/flutter/issues/14375
```
Comment (goderbauer) [inline packages/flutter/test/material/bottom_navigation_bar_test.dart]: Unfortunately, this test still doesn't verify that the traversal order is defined correctly -- and I believe we currently have no way of asserting traversal order in a widget test like this because at this point the SemanticNodes have not been sorted yet?

File: packages/flutter/test/widgets/sliver_semantics_test.dart [hunk 9]
```diff
@@ -151,34 +163,34 @@ void main() {
...
               new TestSemantics(
-                id: 5,
+                id: 6,
                 actions: <SemanticsAction>[SemanticsAction.scrollUp, SemanticsAction.scrollDown],
                 children: <TestSemantics>[
                   new TestSemantics(
                     id: 2,
                     label: r'Item 0',
                     textDirection: TextDirection.ltr,
-                    nextNodeId: 3,
```
Comment (goderbauer) [inline packages/flutter/test/widgets/sliver_semantics_test.dart]: Again, we need to find a way to assert that things are in the correct traversal order in a test.

**Review by goderbauer — COMMENTED** [review]
Some high level comments:

* We should remove the useless `DebugSemanticsDumpOrder.geometricOrder` option and instead replace it with a `DebugSemanticsDumpOrder.traversalOrder` option that allows you to print the semantics tree in, well, traversal order.
* I am still a little afraid of the complexity of this algorithm. I know, this should be mostly mitigated by the fact that n will be small, but still...

File: packages/flutter/lib/src/material/date_picker.dart [hunk 10]
```diff
@@ -641,30 +641,39 @@ class _MonthPickerState extends State<MonthPicker> {
...
         children: <Widget>[
-          new PageView.builder(
-            key: new ValueKey<DateTime>(widget.selectedDate),
-            controller: _dayPickerController,
-            scrollDirection: Axis.horizontal,
-            itemCount: _monthDelta(widget.firstDate, widget.lastDate) + 1,
-            itemBuilder: _buildItems,
-            onPageChanged: _handleMonthPageChanged,
+          new Semantics(
+            sortKey: _MonthPickerSortKey.calendar,
```
[the remark this answers is not in the document]
Comment (yjbanov) [reply packages/flutter/lib/src/material/date_picker.dart]: In general keys are grouped by type, so using plain `OrdinalSortKey` would cause these nodes to be sorted together with potentially unrelated nodes that roll up under the same parent. That's probably not going to happen in the current implementation of the picker.

File: packages/flutter/test/material/date_picker_test.dart [hunk 11]
```diff
@@ -573,43 +579,27 @@ void _tests() {
             ],
           ),
           new TestSemantics(
-            flags: <SemanticsFlag>[
-              SemanticsFlag.isButton,
-              SemanticsFlag.hasEnabledState,
-              SemanticsFlag.isEnabled,
-            ],
+            flags: <SemanticsFlag>[SemanticsFlag.isButton, SemanticsFlag.hasEnabledState, SemanticsFlag.isEnabled],
```
[the remark this answers is not in the document]
Comment (yjbanov) [reply packages/flutter/test/material/date_picker_test.dart]: I used `generateTestSemanticsExpressionForCurrentSemanticsTree` to update this test. I'll fix that method in a separate change when the dust settles.

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 12]
```diff
@@ -1496,92 +1543,271 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
...
-    switch (containerTextDirection) {
-      case TextDirection.rtl:
-        return other.globalStartCorner.dx.compareTo(globalStartCorner.dx);
-      case TextDirection.ltr:
-        return globalStartCorner.dx.compareTo(other.globalStartCorner.dx);
-    }
-    // In case containerTextDirection is null we fall back to child insertion order.
-    return 0;
-  }
+  /// Position within the list of siblings.
```
[the remark this answers is not in the document]
Comment (yjbanov) [reply packages/flutter/lib/src/semantics/semantics.dart]: True. Added.

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 13]
```diff
@@ -1372,6 +1370,55 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
...
+    for (int position = 0; position < childrenInDefaultOrder.length; position += 1) {
+      final SemanticsNode child = childrenInDefaultOrder[position];
+      final SemanticsSortKey sortKey = child.sortKey;
+      lastSortKey = position > 0
+          ? childrenInDefaultOrder[position - 1].sortKey
+          : null;
+      final bool isCompatibleWithPreviousSortKey = position == 0 ||
+          sortKey.runtimeType == lastSortKey.runtimeType &&
+          (sortKey == null || sortKey.name == lastSortKey.name);
+      if (!isCompatibleWithPreviousSortKey && sortNodes.isNotEmpty) {
```
[the remark this answers is not in the document]
Comment (yjbanov) [reply packages/flutter/lib/src/semantics/semantics.dart]: That would be ambiguous as demonstrated by the [B, D, A, C] vs [C, B, D, A] conundrum. I think if your nodes rolled up to a common parent but with different types, chances are that different keys came from different branches of the UI hierarchy, and the most reasonable default is to sort within that hierarchy. If your intention is to sort across branches of the UI, you would provide keys of the matching types.

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 14]
```diff
@@ -1372,6 +1370,55 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
...
+        // Do not sort groups with null sort keys. List.sort does not guarantee
+        // a stable sort order.
```
[the remark this answers is not in the document]
Comment (yjbanov) [reply packages/flutter/lib/src/semantics/semantics.dart]: Correct. I do not expect this to be a problem though. If you are specifying explicit keys, it does not make sense to supply equal keys.

File: packages/flutter/lib/src/semantics/semantics.dart (hunk 1 above)
Comment (yjbanov) [reply packages/flutter/lib/src/semantics/semantics.dart]: I'm not sure how this would work. Only `_TraversalSortNode` knows how to sort by sort key.

File: packages/flutter/lib/src/semantics/semantics.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (yjbanov) [reply packages/flutter/lib/src/semantics/semantics.dart]: Actually, because we are sorting by `_TraversalSortNode` there's no way you can have this situation. `_TraversalSortNode` uses list position in paint order as the tie breaker, so there's no way you can get two equal `_TraversalSortNode`s.

File: packages/flutter/lib/src/semantics/semantics.dart (hunk 2 above)
Comment (yjbanov) [reply packages/flutter/lib/src/semantics/semantics.dart]: Done.

File: packages/flutter/lib/src/semantics/semantics.dart (hunk 4 above)
Comment (yjbanov) [reply packages/flutter/lib/src/semantics/semantics.dart]: Yeah, it's one of those "Are all elements of an empty set blue?" kind of situation. It's just a special case. But I made it clearer in the docs.

File: packages/flutter/lib/src/semantics/semantics.dart (hunk 6 above)
Comment (yjbanov) [reply packages/flutter/lib/src/semantics/semantics.dart]: It uses the position in the list as a tie breaker to guarantee that the sorting is stable. So for now I think this is the most idiomatic method. I also measured the efficiency of the entire algorithm across many of the Flutter Gallery screens. It never exceeded 100µs on Moto G4. This is not surprising as the `SemanticsNode` tree is very sparse compared to the `RenderNode` tree.

File: packages/flutter/test/material/bottom_navigation_bar_test.dart (hunk 8 above)
Comment (yjbanov) [reply packages/flutter/test/material/bottom_navigation_bar_test.dart]: It does now. After https://github.com/flutter/flutter/pull/16253/commits/fad0555668dd95bcf2177dc5762d72a678c85060 `hasSemantics` tests traversal order by default.

File: packages/flutter/test/widgets/sliver_semantics_test.dart (hunk 9 above)
Comment (yjbanov) [reply packages/flutter/test/widgets/sliver_semantics_test.dart]: Done.

File: packages/flutter/lib/src/semantics/semantics.dart (hunk 3 above)
Comment (yjbanov) [reply packages/flutter/lib/src/semantics/semantics.dart]: Done.

File: examples/flutter_gallery/lib/gallery/app.dart [hunk 15]
```diff
@@ -157,21 +157,27 @@ class GalleryAppState extends State<GalleryApp> {
...
+        theme: _galleryTheme.theme.copyWith(platform: _platform ?? defaultTargetPlatform),
+        showPerformanceOverlay: _showPerformanceOverlay,
+        checkerboardRasterCacheImages: _checkerboardRasterCacheImages,
+        checkerboardOffscreenLayers: _checkerboardOffscreenLayers,
+        routes: _kRoutes,
+        home: home,
+        builder: (BuildContext context, Widget child) {
+          return new Semantics(
+            textDirection: _overrideDirection,
+            child: new Directionality(
```
Comment (goderbauer) [inline examples/flutter_gallery/lib/gallery/app.dart]: Directionality is not enough to set the semantic textDirection? Might be worth filing an issue about this.

File: examples/flutter_gallery/lib/gallery/app.dart [hunk 16]
```diff
@@ -42,7 +42,7 @@ class GalleryAppState extends State<GalleryApp> {
   bool _showPerformanceOverlay = false;
   bool _checkerboardRasterCacheImages = false;
   bool _checkerboardOffscreenLayers = false;
-  TextDirection _overrideDirection = TextDirection.ltr;
+  TextDirection _overrideDirection = TextDirection.rtl;
```
Comment (goderbauer) [inline examples/flutter_gallery/lib/gallery/app.dart]: Is this change intentional?

File: packages/flutter/test/widgets/semantics_traversal_test.dart [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+        fail(
+          'Test failed with i == $i, angle == ${angle / math.pi}π\n'
+          '$error'
+        );
+      }
+    }
+  });
+}
+
+class TraversalTester {
```
Comment (goderbauer) [inline packages/flutter/test/widgets/semantics_traversal_test.dart]: Nice!

File: packages/flutter/test/widgets/semantics_traversal_test.dart [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+          expectedTraversal: 1 <= i && i <= 4 ? 'B A' : 'A B',
+        );
+      } catch (error) {
+        fail(
+          'Test failed with i == $i, angle == ${angle / math.pi}π\n'
+          '$error'
+        );
+      }
+    }
+  });
```
Comment (goderbauer) [inline packages/flutter/test/widgets/semantics_traversal_test.dart]: Love all these test cases. Should we add some test cases that use artificial sort keys?

File: examples/flutter_gallery/lib/main.dart [hunk 19]
```diff
@@ -6,6 +6,7 @@
 // Like what you see? Tweet us @flutterio
 
 import 'package:flutter/material.dart';
+import 'package:flutter_gallery/demo/material/date_and_time_picker_demo.dart';
```
Comment (goderbauer) [inline examples/flutter_gallery/lib/main.dart]: revert this file?

File: packages/flutter/test/widgets/semantics_tester.dart [hunk 20]
```diff
@@ -232,7 +214,16 @@ class TestSemantics {
...
 
-  bool _matches(SemanticsNode node, Map<dynamic, dynamic> matchState, { bool ignoreRect: false, bool ignoreTransform: false, bool ignoreId: false }) {
+  bool _matches(
+    SemanticsNode node,
+    Map<dynamic, dynamic> matchState,
+    {
+      bool ignoreRect: false,
+      bool ignoreTransform: false,
+      bool ignoreId: false,
+      DebugSemanticsDumpOrder childOrder: DebugSemanticsDumpOrder.inverseHitTest,
```
Comment (goderbauer) [inline packages/flutter/test/widgets/semantics_tester.dart]: Should this be traversalOrder by default?

File: packages/flutter/test/material/bottom_navigation_bar_test.dart (hunk 8 above)
Comment (goderbauer) [reply packages/flutter/test/material/bottom_navigation_bar_test.dart]: Perfect. 👍

File: examples/flutter_gallery/lib/gallery/app.dart [hunk 21]
```diff
@@ -157,21 +157,27 @@ class GalleryAppState extends State<GalleryApp> {
...
-      checkerboardOffscreenLayers: _checkerboardOffscreenLayers,
-      routes: _kRoutes,
-      home: home,
-      builder: (BuildContext context, Widget child) {
-        return new Directionality(
-          textDirection: _overrideDirection,
-          child: _applyScaleFactor(child),
-        );
-      },
+    return new Semantics(
```
Comment (goderbauer) [inline examples/flutter_gallery/lib/gallery/app.dart]: Is this required for every App? Should it be part of MaterialApp? What changes if I don't specify a top level textDirection for semantics?

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 22]
```diff
@@ -1372,6 +1311,68 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
...
+    TextDirection inheritedTextDirection = textDirection;
+    SemanticsNode ancestor = parent;
+    while (inheritedTextDirection == null && ancestor != null) {
+      inheritedTextDirection = ancestor.textDirection;
+      ancestor = ancestor.parent;
+    }
+
+    List<SemanticsNode> childrenInDefaultOrder = _childrenInDefaultOrder(_children, inheritedTextDirection);
+    if (inheritedTextDirection != null) {
+      childrenInDefaultOrder = _childrenInDefaultOrder(_children, inheritedTextDirection);
```
Comment (goderbauer) [inline packages/flutter/lib/src/semantics/semantics.dart]: Why call `_childrenInDefaultOrder` twice in this case?

File: packages/flutter/lib/src/semantics/semantics.dart (hunk 13 above)
[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/semantics/semantics.dart]: I don't know where the appropriate place is, but it feels like we should document this somewhere.

File: examples/flutter_gallery/lib/gallery/app.dart (hunk 21 above)
Comment (goderbauer) [reply examples/flutter_gallery/lib/gallery/app.dart]: I guess it reverts to paint order?

**Review by goderbauer — COMMENTED** [review]
This looks good! 

Can you update what's left to do in the PR description? I believe it's remove geometric sort order and add hit testing?

File: packages/flutter/lib/src/material/app_bar.dart [hunk 23]
```diff
@@ -473,10 +473,14 @@ class _AppBarState extends State<AppBar> {
       );
     }
 
-    return new Material(
-      color: widget.backgroundColor ?? themeData.primaryColor,
-      elevation: widget.elevation,
-      child: appBar,
+    return new Semantics(
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/app_bar.dart]: Can you add a test that verifies correct traversal order for app bar with a leading icon (e.g. hamburger menu), a title, and a trailing icon (e.g. three dots menu)? That's the traversal order we got most called out for in a11y reviews. Will be good to ensure with tests to keep it correct.

Comment (yjbanov) [reply packages/flutter/lib/src/material/app_bar.dart]: Done.

Comment (yjbanov) [reply packages/flutter/lib/src/material/app_bar.dart]: [direct link](https://github.com/flutter/flutter/pull/16253/commits/c19f8ed851aec31a826e205222625e7fe4fdbf47#diff-47a1cc33e1557fbf135f82b71ac0edeb)

File: packages/flutter/lib/src/semantics/semantics.dart (hunk 22 above)
Comment (yjbanov) [reply packages/flutter/lib/src/semantics/semantics.dart]: My bad, the first one is unnecessary.

File: packages/flutter/lib/src/semantics/semantics.dart [hunk 24]
```diff
@@ -1372,6 +1370,55 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
...
+  List<SemanticsNode> _childrenInTraversalOrder() {
+    final List<SemanticsNode> childrenInDefaultOrder = _childrenInDefaultOrder(_children);
+
+    // List.sort does not guarantee stable sort order. Therefore, children are
+    // first partitioned into groups that have compatible sort keys, i.e. keys
+    // in the same group can be compared to each other. These groups stay in
+    // the same place. Only children within the same group are sorted.
+    final List<_TraversalSortNode> everythingSorted = <_TraversalSortNode>[];
+    final List<_TraversalSortNode> sortNodes = <_TraversalSortNode>[];
+    SemanticsSortKey lastSortKey;
```
[the remark this answers is not in the document]
Comment (yjbanov) [reply packages/flutter/lib/src/semantics/semantics.dart]: A `double` value is undocumentable, so it's unclear what it means to have your nodes sorted together with nodes authored by someone else. The numbers would look magical in the source code. But if the sort keys are provided by _one_ author then they can define and use a custom type.

File: examples/flutter_gallery/lib/gallery/app.dart (hunk 15 above)
Comment (yjbanov) [reply examples/flutter_gallery/lib/gallery/app.dart]: Done. https://github.com/flutter/flutter/issues/16835

File: examples/flutter_gallery/lib/gallery/app.dart (hunk 16 above)
Comment (yjbanov) [reply examples/flutter_gallery/lib/gallery/app.dart]: Oops. Reverted.

File: packages/flutter/test/widgets/semantics_traversal_test.dart (hunk 17 above)
Comment (yjbanov) [reply packages/flutter/test/widgets/semantics_traversal_test.dart]: Thanks!

File: packages/flutter/test/widgets/semantics_traversal_test.dart (hunk 18 above)
Comment (yjbanov) [reply packages/flutter/test/widgets/semantics_traversal_test.dart]: We have them. They just live in one of the `semantics_#_test.dart` files. We should revise this at some point. It's hard to tell what each of those files is testing.

File: examples/flutter_gallery/lib/main.dart (hunk 19 above)
Comment (yjbanov) [reply examples/flutter_gallery/lib/main.dart]: Oops, pulled in my experiments. Fixed.

File: packages/flutter/test/widgets/semantics_tester.dart (hunk 20 above)
Comment (yjbanov) [reply packages/flutter/test/widgets/semantics_tester.dart]: It is now. Thanks for the suggestiong.

**Comment by yjbanov** [comment]
For some reason GitHub is not letting me respond to some of the comments, so responding here:

> ```dart
> isCompatibleWithPreviousSortKey && sortNodes.isNotEmpty
> ```
> I don't know where the appropriate place is, but it feels like we should document this somewhere.

This is documented in `SemanticsSortKey`:

```dart
/// If subclasses of this class compare themselves to another subclass of
/// [SemanticsSortKey], they will compare as "equal" so that keys of the same
/// type are ordered only with respect to one another.
```

I'm not sure this wording is accurate enough though, as it assumes a specific sorting algorithm. I'll try to think of a better wording.

> Can you update what's left to do in the PR description? I believe it's remove geometric sort order and add hit testing?

Done.

File: examples/flutter_gallery/lib/gallery/app.dart (hunk 21 above)
Comment (yjbanov) [reply examples/flutter_gallery/lib/gallery/app.dart]: Reverting for now. Let's address this in https://github.com/flutter/flutter/issues/16835.

**Review by goderbauer — APPROVED** [review]
LGTM
